Compiler support for scala-async, with a state machine transform phase running after erasure [ci: last-only]

retronym opened the pull request:
[`scala-async`](https://github.com/scala/scala-async) ([SIP-22](https://docs.scala-lang.org/sips/async.html)) is currently implemented as a macro.

This pull request incorporates the bulk of the async transform into a fully blown compiler phase of the standard compiler. `scala-async` would remain as a simple macro that expands into the skeleton state machine class including adaptors for `scala.concurrent.Future`.

Other front end macros could also target the async phase adapting to other awaitable data types. The test cases include an [integration with  `java.util.concurrent.CompletableFuture`](https://github.com/scala/scala/pull/8816/files#diff-6fad4e47b4021c5655ebaa6fdfcf45c1R15). It should be also be straightforward to integrate with [`monix.Task`](https://monix.io/docs/2x/eval/task.html) and the like now, unlike [before](https://github.com/scala/scala-async/issues/149).

Why abandon the macro approach?

  - Typer is a somewhat unnatural place in the compiler pipeline for the expansion. The ANF transform is markedly simpler once things after `uncurry` and easier again after `erasure`. Custom integrations of async that use async pattern extactors and guards are _only_  possible when running after `patmat`.
  - Async translated blocks generate larger ASTs than the original source code. Such a a translation is better deferred to later in the pipeline for performance reasons.
  - Async leans heavily on compiler APIs, and some of these are unavailable int the macro API facade.

## Integrate Async into the compiler

  - Import the scala-async code base into `scala.tools.nsc.transform.async`
  - The front end macro is required to expand to `async(expr)(execCtx)` into (approximately) `{val temp = execCtx; class stateMachine { def apply(tr: Try) = { expr; () } }; new stateMachine.start()}`. Wrapping `expr` in a class lets intervening compiler phases do their job (e.g. captured enclosing values give rise to outer pointer usage).
  - The `async` phase proper runs after erasure. As before, it consists of a) a selective ANF transform of the expression; b) a translation of control flow that crosses async boundaries into separate states of a state machine; c) lifting locals that are referred to from multiple states into members of the state machine.
  - Refactor to be idiomatic compiler phase (use normal Transformers/Traversers APIs, use other internal APIs rather than `c.internal_`.
 - This functionality is opt-in with `-Xasync`. A new version of `scala-async` ([prototype](https://github.com/scala/scala-async/pull/237/files)) will be needed to take advantage.
 - Test cases have been imported into `test/async` and `junit/test/scala/tools/nsc/transform/async`.

## Improve the implementation of the ANF transform

  - Refactor the code for better performance and maintainability
  - Be more selective about when to transform code.
  - Simplify with the assumption that only post-erasure Tree shapes will be seen.

## Improve the implementation of the State Machine Transform

  - Refactor the code for better performance and maintainability
  - [Drastically reduce](https://gist.github.com/retronym/50839a051db9b6d1fcd0719d99ecbfc2) the number of states needed to represent pattern matches and if/else constructs with "sparse" usages of await among the branches. Blocks prior to async boundaries are now kept int the same state as the `If` or `LabelDef` itself.
  - Remove the restriction about using `await` in the operand of `Boolean#{||, &&}` and instead rewrite these to `If` trees.
 
## Improve performance

  - Compile time performance is [improved markedly](https://github.com/retronym/compiler-benchmark/pull/1). This is due to emitting fewer states, performing less bookkeeping, and avoiding or fixing inefficiencies in the way we interact with compiler facilities like tree attachments, tree transformers.
  - Runtime performance is likely improved for certain use cases, although the savings may be negligible compared to the cost of async boundaries for many realistic workloads. For each state transition that we avoid, we save one tableswitch, some null field assignments and read/writes of the state variable. Another benefit would be needing to lift fewer values into fields (saving space in the state machine and the put/get field cost.


## TODO
 - [x] Review/Refine the names of the methods in `AsyncStateMachine`
 - [ ] Longer term plan for how to avoid the need for the heuristic approach to deferring `@compileTimeOnly` await methods
 - [x] real world testing against code bases already usnig `scala-async`
 - [x] test coverage for:
   - [x] error cases in `UnsupportedAwaitAnalyzer`
   - [x] `if (tree.tpe =:= definitions.NothingTpe) {` branch in `transformMatchOrIf`
   - [x] `case td@TypeDef(_, _, tparams, rhs)               =>` in `Lifter` (maybe dead code)
   - [x] ThicketTransformer (add a unit test)
   - [x] `case Apply(fun, arg1 :: arg2 :: Nil) if isBooleanShortCircuit(fun.symbol) =>` in `AsyncTraverser`
  - [x] Fix positions of synthetic trees, the position of the apply method (the start of the async block or method) isn't suitable! Either use `NoPosition` of the preciding position. 
  - [ ] Test and update IntelliJ's debugger support for the changes to mangled names: https://github.com/retronym/intellij-scala/pull/1

adriaanm committed:
patch for async debugging

adriaanm committed:
Import async

adriaanm committed:
Implement async transform with a macro and a post-erasure phase

The macro wraps the (untransformed) expression in the skeleton of the
state machine class. Compiler phases between this point and the late
async transform will correctly add outer pointers, lift captured vars
into Ref objects etc. The state machine class itself will also undergo
any transforms that it needs, such as specialization.

The macro is wired up using the FastTrack facility in line with other
macros that are quasi-intrinsic (ie, implemented in the scala-compiler.jar).

After post-erasure, we run the ANF and state machine transform.

The old code used the macro refletion API -- these usages are
now being replaced with direct use of the richer Global API.

adriaanm committed:
make xprompt work in sbt's scalac task

retronym committed:
Improvements to async phase

Passes the correct owner symbol to `useFields` to fix a bug
with local modules.

Refactors to store state about the current async block in a var
which allows us to mix in all the implementation once into
AsyncPhase.

Remove some testing FutureSystems. We'll refactor the tests
to use ScalaConcurrentFutureSystem.

Allow a custom front end to supply an alternative FutureSystem
attached to the block defining the state machine wrapper class.
Demonstrate this with a test case for an annotation driven
front end that targets an alternative Future implementation.

Remove some now-unused customization points from FutureSystem.
For example, we don't need to abstract over the choice of
base class for the state machine as a custom front end now
creates the state machine tree directly.

FutureSystem now deals directly in Trees, rather than in
`c.Expr`. I've kept a light weight option (a type alias
`type Expr[T] = Tree`) so we can have a little compiler support
for getting the types right and documentation for
`FutureSystem` implementors.

Refactor the ANF transform (and others) to directly be a
TypingTransformer.

Remove the need to subclass Function0 in the state machine
by kicking things off with `Future.unit.onComplete(stateMachine)`.
This reduces bloat in the generated code due to Function0
specialization.

I've also elminated the macro application in the early transform
in favour of wrapping the argument in `Block(<arg> :: Nil, q"()"}`.
This ran into some "pure expression has no effect" warnings which
I've suppressed with a new tree annotation.

Most of the `auto_` test cases are converted to use the
standard scala.concurrent version of async. Now that the
transform is always happening post-erasure, there is no need to
keep those tests under the annotation driven async test (which
used to be the only way to get post-typer expansion).

A few test can only be expressed with the annotation driven
version of async, namely those that have an async boundary
in an extractor. These have been moved into the
`AnnotationDrivenAsyncTest`.

szeiger committed:
Move to scala.async.Async for compatibility with existing macros

szeiger committed:
Add JUnit support to partest

szeiger committed:
Add tests from scala-async

  - most regular scala.async.Async tests pass
  - AsyncId converted to ScalaConcurrentAsync
  - non-standard tests not properly integrated yet
  - actual test failure in `run/futures.scala`
  - neg/NakedAwait still missing (alternative first)
  - remove late.scala (converted to AnnotationDrivenAsync)
  - Remove uncheckedBounds.scala (the problem no longer exists post-erasure)

retronym committed:
Convert tests to ScalaConcurrent Future System

Also:
  - remove late.scala (converted to AnnotationDrivenAsync)
  - Remove uncheckedBounds.scala (the problem no longer exists post-erasure)

retronym committed:
Fix nested async transforms

UseFields needs to do what explicit outer would have done:
access inner classes must access fields via the outer
parameter and we must ensure fields accessed from inners
are not JVM private.

retronym committed:
A simpler and faster ANF transform

A long overdue overhaul.

Originally, we defined this in terms of a pair of mutually recursive
transforms (`anf` and `linearize`). We can simplfy the implementation
by collapsing these the cases of single Transformer.

This transformer is designed for "thicket" based transforms, where
a single input tree can map to a list of output trees which will
be flattened into the enclosing `Block`. An enclosing block will be
automatically added if there was none before.

Make the transform more idiomatic by more widespread use of
utilities to generated attributed trees rather than passing
untyped trees through `localTyper.typed`.

Rework special casing of Unit/Nothing typed expressions to
avoid temporary vals of these types to avoid littering the
resulting trees with `BoxedUnit` references.

Avoid unnecessary temporaries in:
  - the suffix of the args of an Apply the follows the final `await`
  - the result of If that only awaits the condition
  - similarly, the result of a Match that only awaits the scrutinee
  - "safe to inline" expressions like simple Idents or constants.

Remove special cases for derived value classes which are
eliminated now before the ANF transform.

Furthermore:

  - Deal with ArrayValue tree
  - Remove cast which was used ot suppress typer warning
  - Don't need to deal with nested Applys after uncurry
  - Remove an new-unneeded adjustment from the entry point to the ANF
    and from MatchResultTransformer.

retronym committed:
Deal with await in ArrayValue

As a late-running ANF transform must do.

retronym committed:
Companion Detection should be post-erasure aware

Deal with the representation of local modules at async's
new place in the compiler pipeline.

Use more efficient/idiomatic collection of local def trees.
(`Tree#children` creates temporary lists.)

retronym committed:
Reduce allocations in name freshening.

retronym committed:
Live variables: detect usages in Awaitable.expr

retronym committed:
Fix broken attempt to elide nulling in final state

retronym committed:
Sort cases by state ID

retronym committed:
Fix .dot output when code contains ampersand

retronym committed:
We'll never see `isAsync` again in nested trees.

This code path was only relevant when async was run as a macro
in the presentation compiler.

retronym committed:
Use reporter.error directly

retronym committed:
Cleanup AsyncPhase's transformer

Transform the ClassDef and DefDef in separate cases.
This reduces some of manual atOwner and treeCopy busywork.

retronym committed:
Harden Context.make debug logging against null trees.

TypingTransformers can easily feed null trees into here
if they don't assign `curTree`.

retronym committed:
Integrate async's logging into debuglog

The ANF transform trace is demoted to the
"edit a boolean and re-compile" UI.

retronym committed:
Refactor and document UseFields transform

Also, lifted lazy var fields need not be MUTABLE.

retronym committed:
Reduce number of states in async state machine

We used to lift the control flow for any pattern match or
if/else into fine grained states if it enclosed even a single
await call.

Instead, we now keep as much of the control flow as possible
in the current state. The program point after the control flow
statement is lifted into a new state iff it is reachable from
any branch that contains an await.

LabelDef-s representing cases and the matchEnd of patterns are now
only lifted into states if they are the target of a label jump
from a resumption handler.

To make this analysis possible, we form a context chain of
AsyncBlockBuilders to let the translation of the label jumps compare
the current state to that current state of the enclosing block builder
that holds the LabelDefs. If these differ, we can conclude that we
have crossed an async boundary and mark the LabelDef for lifting
into state.

Here's how the simplification helps:

https://gist.github.com/retronym/50839a051db9b6d1fcd0719d99ecbfc2

Refactorings along the way:

  - introduce a ThicketTransformer to transform `Tree => List[Tree]`
  - make `symLookup` more convenient to access
  - add some internal flags to disable state compaction and field
    nulling which help to debug the phase
  - unify onto a single `AsyncState` class. Different use cases
    are served by eagerly adding additional logic, rather than
    having this added later on in the variants of
    `mkCompletionHandler`
  - detect and remove empty states in `compactStates`. It seems easier
    now to do this than to avoid constructing them.
  - Try a bit harder to avoid redundant `()` in stats.
  - Move synthesis of async block generation into
    StateBuilder.build (formrly known as resultSimple). This is
    now the only factory method on StateBuilder.
  - Fix propagation of the enclosing subsequent state into
    nested expression positions
  - In combination, this lets us perform async block completion
    inline in all terminal states, there is no need to write the
    result to lifted field and make a state transition.
  - Remove the `currTree` var in favour of delegating to
    the field in the `stateBuilder` var
  - Encapsulate state transition bookkeeping

retronym committed:
Avoid overhead in tree attachments

Deal with the underlying Set directly rather than
using the ClassTag indexed lookup/removal.

retronym committed:
Use faster attachment lookups in ThicketTransformer

retronym committed:
Fix bug in StateSet and add a remove operation

retronym committed:
Reinstate live variables test and fix null assignment

While looking more closely at the implementation, I
realised our approach of nulling dead variables at the
start of the subsequent state was wrong, we retained
the references for longer than needed.

I've moved the nulling into the conclusion of the
state of last usages.

Remove the warnings test which isn't relevant with post-typer ANF.

retronym committed:
Delete dead code.

retronym committed:
Use compiler APIs.

retronym committed:
Deal in typed trees in ExprBuilder.

retronym committed:
Remove PhasedTransform

We don't have any phase-dependent logic left.

retronym committed:
Remove legacy future system methods.

retronym committed:
Remove unused imports

retronym committed:
Allow a nested TypingTransformer to use an existing local typer.

This saves allocations for root imports etc.

retronym committed:
Look for references in types as well as in terms.

retronym committed:
Merge AsyncTransformState with SymLookup

Also:
 - remove some redundant or simply unused parameters.
 - tighten access on some members

retronym committed:
Add a fast path to the async phase

If no methods have been marked for tranform, disable
the phase for the entire unit.

retronym committed:
Split some async implementation classes into files

retronym committed:
Refactor lifting to to change owner sooner

And reuse existing utility method to change module class owner.

retronym committed:
Refactor field nulling.

retronym committed:
Remove unused imports.

retronym committed:
Remove dead cases from adaptToUnit

The input trees are always typed now which means the early cases
subsume the ones being deleted here.

retronym committed:
Move AsyncTransform code into AsyncPhase.scala

retronym committed:
Use of AnfTransformer directly.

retronym committed:
Refactor asyncTransform

retronym committed:
Refactor addFutureSystemAttachment

retronym committed:
Remove dead code.

retronym committed:
Avoid allocating containsAwaitTraverser so often.

retronym committed:
Avoid logging string allocation

retronym committed:
Comment out debugging toString.

retronym committed:
More efficient Contains-/No-Await cleanup

retronym committed:
Avoid overload resolution in typing default case of state handler.

retronym committed:
Limit UseFields transform to relevent subtrees.

retronym committed:
More typed sub-trees avoiding working in `typed(...)`

retronym committed:
Inline case class Awaitable

retronym committed:
Overhaul integration with front end macros

  - The state machine generated by the front end macro must Now
    implement a small set of methods as a facade over the
    particular Future/Awaitable/Task type being used. This
    replaces the AST factory methods in `FutureSystem`
  - Remove intrinsic implemnentation for scala-async's async
    macro. Show what it will need to do in a test implementation,
    `s.t.n.partest.Async` and update all test cases to use this.
  - Add a heuristic to interpret `@compileTimeOnly` annotated
    methods to register `await` methods. Defer the check in refchecks
    for all of these and instead check that no references survive
    the async phase.
  - Refactor the test implementations of async front ends to
    share an interface for the state machine. We don't ship
    this but it could be copy/pasted into third party integrations.
  - Add a test integration with Java's CompletableFuture.
  - Expose a method through macro `Internals` API to let
    front ends mark a method in the state machine for async
    translation.

retronym committed:
Be explicit about lambdaLift following the async phase.

retronym committed:
StateSet: Allocate caseSet lazily

sjrd commented:
Can these changes create new shapes of trees using `LabelDef`s, notably in the interaction with pattern matching? That would potentially mean more work on Scala.js' side.

retronym commented:
@sjrd Yes, I'm using label jumps like `continue` 

```
def apply(tr: Try[AnyRef]) = {
:while1
while (true) {
  state match {
     case <i> => 
        ...
          GOTO while1();
        ...
           GOTO await(j, future)
     case _ => throw ...
  }
  GOTO while1
  :await(newState, awaitable)
   state = newState
   if (isComplete(awaitable) tr = awaitable.get else {awaitable.onComplete(this); return}
}
}
```

The `:await` label could be inlined into its call sites by either Scala.js post processing or by the async phase when it detects Scala.js is active. It isn't essential to the control flow, it is an optimization to reduce code size.

The `GOTO while1` would need to be translated to a Javascript `continue`.

sjrd commented:
We don't compile to JavaScript, but to our IR. Our IR doesn't have `continue`, but it has something much more fun: labeled blocks (see section 3.2.6 of [my thesis](https://lampwww.epfl.ch/~doeraene/thesis/)). From what you're showing, it seems we'll definitely need some more special casing to recognize that shape, but that we can translate it to the following IR:
```scala
while (true) {
  while1[void]: {
    val (newState, awaitable) = await[(int, Awaitable)]: {
      state match {
        case <i> =>
          ...
          return@while1 void
          ...
          return@await (j, future)
        case _ => ...
      }
      return@while1 void
    } // end await
    state = newState
    if (isComplete(awaitable))
      tr = awaitable.get
    else
      {awaitable.onComplete(this); return}
  } // end while1
}
```

sjrd commented:
It seems I replied to an earlier version of your comment.

For `case` labels, we already have a solution, as long as their shape doesn't get more complicated because of the transform.

retronym committed:
Don't insert null assignments for the terminal state.

retronym committed:
Remove dead code or make things more private.

retronym committed:
Local TypeDef no longer make it to the async phase.

retronym committed:
Allow compilation of async's partest suite in Junit

This lets us use IntelliJ's code coverage analysis.

retronym committed:
Test and simplify ANF transform of Nothing-typed if/match

retronym committed:
Fix detection of ill-nested awaits

 - Adapt detection logic to the post-erasure tree shapes
 - Lift restriction about || and && in favour of a
   rewrite to `If` in the ANF transform
 - import relevant parts of the test from scala-async.

retronym committed:
Move ThicketTransformer to s.t.n.transform

It seems generally useful enough to move it out of
the async utilities. I've also added a unit test.

retronym committed:
Add tests from scala/scala-async#210

https://github.com/scala/scala-async/pull/210

retronym committed:
Add tests from scala/scala-async#206

https://github.com/scala/scala-async/pull/206

retronym committed:
Fix regression with non-awaiting patterns in tail position

retronym committed:
More granular fresh names.

retronym committed:
erasure: push casts into LabelDefs to simplify async phase

retronym committed:
Avoid VerifyErrors wrapping Try in Assign.

retronym committed:
await in by-name function arg is repoted as ill-nested.

retronym committed:
Overhaul/fix live variables

retronym committed:
Improve error reporting in asynj junit test.

Show source file context for reported compiler
errors and warnings, as we would expect from
the normal ConsoleReporter.

sjrd commented:
I'd like to discuss how we can make this testable in Scala.js. At the moment it seems there are essentially two kinds of tests:
* partest tests that actually delegate to a kind of JUnit runner, with tests that systematically perform `Await.result`
* "driven" tests relying on a JUnit test that instantiates a compiler, then runs the compiled code via reflection

Neither kind will be testable in Scala.js, which becomes problematic now that the support of `async` can generate new shapes of trees. It means that we'll never be able to reliably test any of this in the context of Scala.js.

Could we instead use one way to write async tests that is (or can be made) compatible with Scala.js? Off the top of my head, I imagine we could have a single `AsyncTest` helper in `partest-extras`, that would like:
```scala
abstract class AsyncTest {
  def test: Future[Any]

  def main(args: Array[String]): Unit {
    Await.result(test, Duration.Inf)
  }
}
```
Then, *all* tests that need running would be written as partests that inherit from this class. In Scala.js, we would override that single class so that it doesn't use `Await.result`, but instead delegates to the event loop. That would be manageable, and would allow all these tests to be testable also on Scala.js.

retronym committed:
Support state getter/setter methods

Rather than assuming its a var.

retronym committed:
Better integration with compiler fresh names

 - No longer add the `$async` suffix to fresh names. Add a test to
   show that we no longer risk clashing with lambda lifted names
   because we're coordinating through the unit's FreshNameCreator.
 - Add a facility to FreshNameCreator to support clients that
   want to generate many fresh names from a single prefix
   and avoid the overhead of going through `newTermName`
   when subsequent units also want to generate `x$0`, etc.
 - Add a little efficiency to name concatenation by use of
   presized of string builders.

retronym commented:
@sjrd That sounds like a reasonable plan. 

The JUnit tests are written in that way because either:

   a) they need to test the compiler plugin front-end that use `@autoasync`. However I could move this compiler plugin implementation into `partest-extras` and activate the plugin in selected partests with `-Xplugin`.
   b) I added the tests there because it was more convenient for me to debug that an equivalent `partest` case. I can just move those back.

The partest cases can certainly be refactored to abstract over the `Await.result` call.

retronym committed:
Position synthetic trees accurately

retronym committed:
Add an another example of a Writer/Option monad

retronym commented:
/rebuild

retronym commented:
@sjrd I've converted one of the partest Junit tests to a reflection free (other than a `ClassTag) harness that doesn't directly use `Await.result`. Here's how it looks: https://github.com/scala/scala/compare/a29adccbbcea481abce38249dc54eda0b6baae72...retronym:topic/async-scalajs-test-compat?expand=1

Does this look okay?

sjrd commented:
Thank you @retronym. It's still problematic because there is no way to implement
```scala
def block[T](f: Future[T]): T
```
on the JS platform.

Essentially each test must return a `Future[Unit]` instead of a `Unit`, and it must be the responsibility of `main` to to block on those futures (on the JVM) or to chain them through the even loop (in JS).

The method `test` would have the following signature:
```scala
  private val tests = mutable.ListBuffer[(String, () => Future[Unit])]()
  def test(name: String)(f: => Future[Unit]) = {
    tests += (name, () => f)
  }
```
and for example, the test that is currently:
```scala
    test("inlining block does not produce duplicate definition") {
      async {
        val f = 12
        val x = await(fut(f))

        {
          type X = Int
          val x: X = 42
          identity(x)
        }
        type X = Int
        x: X
      }.block
    }
```
should would lose the `.block` at the end.

A test that performs assertions after the `.block`, like
```scala
    test("inlining block in tail position does not produce duplicate definition") {
      async {
        val f = 12
        val x = await(fut(f))

        {
          val x = 42
          x
        }
      }.block mustBe (42)
```
would instead end in
```scala
      }.map { _ mustBe (42) }
```

retronym committed:
Fixup remove stray .log files from partest suite

retronym committed:
Move async run tests to jvm category

Some will be made Scala.js compatible in a subsequent commit and
moved back to run.

retronym committed:
Add a Scala.js compatible smoke test.

retronym committed:
Move some async test cases to partest

retronym commented:
@sjrd I've taken a middle ground by moving a subset of the test cases to use a synchronous DSL over `Option` rather than the asynchronous DSL over `s.c.Future`. These tests also don't use JUnit's reflection. The other tests are now segregated in `test/async/jvm`.

sjrd commented on src/partest-extras/scala/tools/partest/async/Async.scala:
```diff
@@ -0,0 +1,80 @@
+    def mark(t: DefDef): Tree = {
+      c.internal.markForAsyncTransform(c.internal.enclosingOwner, t, awaitSym, Map.empty)
+    }
+    val name = TypeName("stateMachine$async_" + body.pos.line)
```
Could we avoid the `_` sign in the class name, please? That triggers a bail-out in the Scala.js back-end, which requires the use of Unicode characters in the generated JS identifiers. It works, of course, but it's inefficient (while generating the code) and ugly (in the generated .js file).

sjrd commented on src/partest-extras/scala/tools/partest/async/AsyncStateMachine.scala:
```diff
@@ -0,0 +1,35 @@
+
+// The async phase expects the state machine class to structurally conform to this interface.
+trait AsyncStateMachine[F, R] {
+  /** Assign `i` to the state variable */
+  protected def state$async_=(i: Int): Unit
+  /** Retrieve the current value of the state variable */
+  protected def state$async: Int
```
Could we not use `$` in method names that users will have to see at some point? `$` are supposed to be reserved for the compiler, so we shouldn't encourage people by showing them us using such `$`s.

sjrd commented:
@retronym Thank you. I managed to update the Scala.js backend enough to support the `tests/async/smoketest.scala`. See https://github.com/scala-js/scala-js/pull/4067 for the changes.

You said earlier:

> The `GOTO matchEnd` or `GOTO caseNext` labels now can target a label that is not in the same `Block`, which I expect would be problematic for your translation. These could be replaced with an extra pass (either by the async phase or by Scala.js) with `state = <state containing label>; while1()`.

I did not see any occurrence of this pattern in the smoke test. Consequently my updated Scala.js backend still does not support that. Is it really true that it can happen? If yes, could you give an example source code, ideally with `optionally`, that triggers this?

retronym committed:
Tweak names in generated code

 - Remove '_' and line number from state machine classs names.
 - Rename `state$async` to `state`.

retronym commented:
> I did not see any occurrence of this pattern in the smoke test. 

This PR no longer emits this AST shape, and I don't plan to use it now.

sjrd commented:
I see. I could indeed reproduce that behavior.

The bad news is: I have no idea how to deal with this in the Scala.js back-end. If you look at the PR I mentioned above, https://github.com/scala-js/scala-js/pull/4067, and in particular at the commit "Generalize the codegen of `LabelDef`s for pattern matching", you can see the places where with deal with patmat-generated `LabelDef`s. You'll see that with the changes of that PR, we can in general deal with the following shapes:
```scala
  {
    prologue
    case1() {
      ...
        matchEnd(1)
      ...
        case2()
    }
    case2() {
      ...
    }
    ...
    caseN() {
      ...
    }
    <optional>:
      matchEnd(x) {
        ...
      }
    rest
  }
```
which we translate as:
```scala
// Scala.js IR
val x = matchEnd[int]: {
  case2[void]: {
    ...
      return@matchEnd 1
    ...
      return@case2 undefined
  }
  case3[void]: {
    ...
  }
  caseN[void]: {
    ...
  }
  ... // content of caseN() before
}
x
```
We can do this because all jumps are at least to a label that is defined in an enclosing scope of where the jump happens.

Now with the new async transformation, it seems we can have a jump in one scope calling a LabelDef defined in an unrelated scope. It is very problematic for Scala.js, but honestly I also find that a bit terrifying in terms of sanity of scalac Trees.

You suggested that instead of having a jump to `matchEnd()`, we could have something like
```scala
async_70$1.this.match$2 = 2;
async_70$1.this.state$async_=(3);
while$()
```
Recovering that after the fact seems like a daunting task. How would we identify those places? How would we make sure that we only alter async-generated stuff, and not other code? It seems completely unwieldy.

**Couldn't the async transform directly generate the above?** That would avoid having jumps to labels that are not declared in the scope of the jumps, which seems like a decent property to have, even regardless of Scala.js.

The only concern I can think of is that it would generate less-than-optimal bytecode on the JVM. But that seems to already happen in other places, for example in the `case 2 =>` above.

It seems to me that this would be the best compromise between the requirements of the two platforms.

Another reason is that the schema wouldn't be portable to Dotty. Dotty has exactly the same notion of labeled blocks as the Scala.js IR, so we would have to generate something compatible with those `Labeled` blocks anyway.

retronym committed:
Avoid exotic use of Labels to gain Scala.js compat

retronym commented:
I've added a post-processing pass to the `async` phase to find residual label jumps that cross state transitions and replace them with `state = ...; while()`.

sjrd commented:
> I've added a post-processing pass to the async phase to find residual label jumps that cross state transitions and replace them with state = ...; while().

Thank you! This is perfect. The snippet from above worked out of the box with your latest changes. :)

Thanks also for addressing my other comments :)

retronym committed:
Add missing source file headers

retronym committed:
Fix an edge case for value classes

retronym committed:
Fix an edge case for while in expr position

lrytz commented on src/reflect/scala/reflect/api/Internals.scala:
```diff
@@ -374,6 +374,8 @@ trait Internals { self: Universe =>
      */
     def boundedWildcardType(bounds: TypeBounds): BoundedWildcardType
 
+    def markForAsyncTransform(owner: Symbol, method: DefDef, awaitSymbol: Symbol, config: Map[String, AnyRef]): DefDef = method
```
Would be good to have some comment here; this is the entry point for defining custom async-style macros, right? e.g. https://github.com/retronym/monad-ui/blob/master/src/main/scala/monadui/OutputAsyncDsl.scala

retronym replied:
Comments added.

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -568,11 +575,17 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
+  object async extends {
+    val global: Global.this.type = Global.this
+    val runsAfter = List("posterasure")
+    val runsRightAfter = Some("posterasure")
```
is the hard link here required, or coud it be `None`?

retronym replied:
`None` seems fine as `lambdalift` asks that it `runsAfter` `async`.

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -1336,6 +1350,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     def runIsAt(ph: Phase)   = globalPhase.id == ph.id
     def runIsAtOptimiz       = runIsAt(jvmPhase)
 
+    firstPhase.iterator.foreach(_.init())
```
Leftover? It seems there's no phase that does something in `init()`.

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -999,7 +1013,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
     && rootMirror.isMirrorInitialized
   )
-  override def isPastTyper = isPast(currentRun.typerPhase)
+  override def isPastTyper = globalPhase != null && isPast(currentRun.typerPhase)
```
I don't find any assignments of `null` to `globalPhase`?
```
  var globalPhase: Phase = NoPhase
```

retronym replied:
Removed, this must have come up when I was trying the `Phase.init()` hook.

lrytz committed:
remove old ThicketTransformer

sjrd commented:
With a few more changes on Scala.js' side, I could make all the tests in edge-cases pass :)

I noticed that, with the latest state of this PR, (almost) all generated switch'es have a last state that is empty dead code, and that would loop forever if reached. For example, the very first `testBasic()` in `smoketest.scala` generates:
```scala
  final class Test$stateMachine$async$1 extends scala.tools.partest.async.OptionStateMachine {
    <synthetic> <stable> private[this] var await$1: Object = _;
    override def apply(tr$async: Option): Unit = while$(){
      try {
        Test$stateMachine$async$1.this.state() match {
          case 0 => {
            val awaitable$async: Some = new Some(scala.Int.box(1));
            tr$async = Test$stateMachine$async$1.this.getCompleted(awaitable$async);
            Test$stateMachine$async$1.this.state_=(1);
            if (null.!=(tr$async))
              while$()
            else
              {
                Test$stateMachine$async$1.this.onComplete(awaitable$async);
                return ()
              }
          }
          case 1 => {
            {
              val tryGetResult$async: Object = Test$stateMachine$async$1.this.tryGet(tr$async);
              if (Test$stateMachine$async$1.this.==(tryGetResult$async))
                Test$stateMachine$async$1.this.await$1 = return ()
              else
                Test$stateMachine$async$1.this.await$1 = tryGetResult$async.$asInstanceOf[Object]()
            };
            val awaitable$async: Some = new Some(scala.Int.box(2));
            tr$async = Test$stateMachine$async$1.this.getCompleted(awaitable$async);
            Test$stateMachine$async$1.this.state_=(2);
            if (null.!=(tr$async))
              while$()
            else
              {
                Test$stateMachine$async$1.this.onComplete(awaitable$async);
                return ()
              }
          }
          case 2 => {
            <synthetic> val await$2: Object = {
              val tryGetResult$async: Object = Test$stateMachine$async$1.this.tryGet(tr$async);
              if (Test$stateMachine$async$1.this.==(tryGetResult$async))
                return ()
              else
                tryGetResult$async.$asInstanceOf[Object]()
            };
            <synthetic> val x$1: Int = scala.Int.unbox(await$2);
            <synthetic> val x$2: Int = scala.Int.unbox(Test$stateMachine$async$1.this.await$1).+(x$1);
            Test$stateMachine$async$1.this.completeSuccess(scala.Int.box(x$2));
            return ();
            Test$stateMachine$async$1.this.await$1 = null
          }
          case 3 => {
            ()
          }
          case _ => throw new IllegalStateException(java.lang.String.valueOf(Test$stateMachine$async$1.this.state()))
        }
      } catch {
        case (throwable$async @ (_: Throwable)) => {
          Test$stateMachine$async$1.this.completeFailure(throwable$async);
          return ()
        }
      };
      while$()
    };
    def <init>(): Test$stateMachine$async$1 = {
      Test$stateMachine$async$1.super.<init>();
      ()
    }
  }
```
Note the `case 3 => { () }`. If reached, it would loop forever since it doesn't change the state, but no other code sets the state to 3. The same happens in all async blocks I looked at.

This is not wrong/harmful; it's just useless trees.

retronym committed:
Rework async expression wrapping

Wrap the async expression in `locally { ... }` to
force value class boxing. This seems to work better
than `{ ... }: Any`, which ends up with `Function`
trees typed with `Any` which violoates an assumption
in `delambdafy`.

retronym committed:
Avoid generating dead code for terminal state

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -327,10 +327,15 @@ trait Implicits {
-    name match {
-      case x: TermName => result.typeSymbol.newMethod(x) setInfoAndEnter tp
-      case x: TypeName => result.typeSymbol.newAbstractType(x) setInfoAndEnter tp
+    val owner = result.typeSymbol orElse { // TODO async (when after erasure?)...
```
any insight why implicit search triggers for code generated by async?

retronym replied:
Doesn't seem to happen anymore, probably a leftover from the initial state of the PR in 6a665e97f991c1fe039fdb269c5b26399967ccaa or an even earlier version. Removed

lrytz commented on src/compiler/scala/tools/reflect/FastTrack.scala:
```diff
@@ -44,6 +44,9 @@ class FastTrack[MacrosAndAnalyzer <: Macros with Analyzer](val macros: MacrosAnd
   private def makeWhitebox(sym: Symbol)(pf: PartialFunction[Applied, MacroContext => Tree]) =
     sym -> new FastTrackEntry(pf, isBlackbox = false)
 
+  private def makeBlackBoxIfExists(sym_pf: (Symbol, PartialFunction[Applied, MacroContext => Tree])) =
```
unused?

lrytz commented on src/reflect/scala/reflect/internal/StdAttachments.scala:
```diff
@@ -119,4 +119,9 @@ trait StdAttachments {
+  // When typing a Def with this attachment, change the owner of its RHS from origalOwner to the symbol of the Def
+  case class ChangeOwnerAttachment(originalOwner: Symbol)
+
+  case object SuppressPureExpressionWarning extends PlainAttachment
```
Unused?

lrytz commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -478,7 +478,14 @@ trait Trees extends api.Trees {
+
+    override def tpe_=(tp: Type): Unit = {
+      if (!definitions.isFunctionType(tp))
+        new Throwable().printStackTrace()
```
remove debug code?

lrytz commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -1531,6 +1538,7 @@ trait Trees extends api.Trees {
   }
 
   class ChangeOwnerTraverser(val oldowner: Symbol, val newowner: Symbol) extends Traverser {
+    assert(newowner != NoSymbol, oldowner)
```
remove, or extend message?

lrytz reviewed (commented):
A few minor comments. I also pushed a commit to remove (hopefully) unused code.

retronym committed:
Address review comments (remove dead/debug code, +comments)

retronym committed:
Refactor detection of ill nested await calls

We can defer the check until the current location of the "fallback"
check without sacrificing the specific error messages.

lrytz commented:
About the risk for regressions that this change may cause:
  - All non-trivial changes in this PR are behind a new `-Xasync` compiler flag (for now)
  - Existing scala-async releases will continue to work exactly the same way with compilers that include this change
